File: fridge-service/src/main/java/com/zenika/tutoriels/observabilitydemo/fridge/business/Fridge.java
package com.zenika.tutoriels.observabilitydemo.fridge.business;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

import static java.lang.Math.random;

public class Fridge {

    private static final Logger LOGGER = LoggerFactory.getLogger(Fridge.class);

    private final Map<FoodStuff, Map<String, AtomicInteger>> stocks;

    private final FailingService failingService;


    public Fridge(MeatInitialStock meatInitialStock, CheeseInitialStock cheeseInitialStock, FailingService failingService) {
        this.stocks = Map.of(FoodStuff.meat,
                Map.of("chicken", new AtomicInteger(meatInitialStock.chicken()),
                        "beef", new AtomicInteger(meatInitialStock.beef()),
                        "pork", new AtomicInteger(meatInitialStock.pork())),
                FoodStuff.cheese,
                Map.of("cheddar", new AtomicInteger(cheeseInitialStock.cheddar()),
                        "goat", new AtomicInteger(cheeseInitialStock.goat()),
                        "blue", new AtomicInteger(cheeseInitialStock.blue())
                )
        );
        this.failingService = failingService;
    }

    public List<String> listAvailableMeats() {
        return getStock(FoodStuff.meat);
    }

    public List<String> listAvailableCheeses() {
        return getStock(FoodStuff.cheese);
    }

    public void removeMeatFromFridge(String meat) {
        removeFromStock(FoodStuff.meat, meat);
    }

    public void removeCheeseFromFridge(String cheese) {
        removeFromStock(FoodStuff.cheese, cheese);
    }

    private List<String> getStock(FoodStuff foodStuff) {
        LOGGER.debug("Listing available {}s", foodStuff.name());
        return stocks.get(foodStuff).entrySet().stream()
                .filter(entry -> entry.getValue().get() > 0)
                .map(Map.Entry::getKey)
                .collect(Collectors.toList());
    }

    private void removeFromStock(FoodStuff stuff, String kind) {
        LOGGER.debug("Removing a porting of {} from {} stock", kind, stuff);
        this.failingService.potentiallyFail();
        AtomicInteger capacity = this.stocks.get(stuff).get(kind);
        if (capacity == null) {
            throw new UnknownFoodStuffException(stuff, kind);
        } else if (capacity.decrementAndGet() < 0) {
            LOGGER.error("No more {}: {}", stuff, kind);
            throw new NoMoreFoodStuffException(stuff, kind);
        }
    }

}

File: fridge-service/src/main/kotlin/com/zenika/tutoriels/observabilitydemo/fridge/business/Fridge.kt
package com.zenika.tutoriels.observabilitydemo.fridge.business

import com.zenika.tutoriels.observabilitydemo.logger.utils.ServiceLoggerAutoConfiguration.Companion.withRequestInfoInMdc
import com.zenika.tutoriels.observabilitydemo.logger.utils.ServiceLoggerAutoConfiguration.Companion.withRequestInfoInMdcForLog
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import reactor.core.publisher.Mono
import java.lang.Math.random
import java.util.*
import java.util.concurrent.atomic.AtomicInteger
import java.util.stream.Collectors

enum class FoodStuff {
    cheese, meat
}


data class MeatInitialStock(val chicken: Int, val beef: Int, val pork: Int)

data class CheeseInitialStock(val goat: Int, val cheddar: Int, val blue: Int)

class NoMoreFoodStuffException(val foodStuff: FoodStuff, val kind: String) : RuntimeException()

class UnknownFoodStuffException(val foodStuff: FoodStuff, val kind: String) : RuntimeException()

class RandomException: RuntimeException()


class Fridge(meatInitialStock: MeatInitialStock, cheeseInitialStock: CheeseInitialStock) {

    companion object {
        private val LOGGER: Logger = LoggerFactory.getLogger(Fridge::class.java)
    }

    private val stocks: EnumMap<FoodStuff, Map<String, AtomicInteger>> = EnumMap(FoodStuff::class.java)

    init {
        this.stocks[FoodStuff.meat] = mapOf("chicken" to AtomicInteger(meatInitialStock.chicken), "beef" to AtomicInteger(meatInitialStock.beef), "pork" to AtomicInteger(meatInitialStock.pork))
        this.stocks[FoodStuff.cheese] = mapOf("cheddar" to AtomicInteger(cheeseInitialStock.cheddar), "goat" to AtomicInteger(cheeseInitialStock.goat), "blue" to AtomicInteger(cheeseInitialStock.blue))
    }

    fun removeMeatFromFridge(meat: String): Mono<Void> {
        return removeFromStock(FoodStuff.meat, meat)
    }

    fun removeCheeseFromFridge(cheese: String): Mono<Void> {
        return removeFromStock(FoodStuff.cheese, cheese)
    }

    fun listAvailable(foodStuff: FoodStuff): Mono<List<String>> {
        return Mono.just(foodStuff)
                .doOnEach(withRequestInfoInMdcForLog { LOGGER.debug("Listing available {}s", foodStuff.name) })
                .map { s ->
                    stocks[s]!!.entries.stream()
                            .filter { it.value.get() > 0 }
                            .map { it.key }
                            .collect(Collectors.toList())
                }
    }

    private fun removeFromStock(stuff: FoodStuff, kind: String): Mono<Void> {
        return Mono.zip(Mono.just(stuff), Mono.just(kind))
                .doOnEach(withRequestInfoInMdcForLog { LOGGER.debug("Removing a porting of {} from {} stock", it!!.t2, it.t1) })
                .flatMap(withRequestInfoInMdc { remove(it.t1, it.t2) })
    }

    private fun remove(stuff: FoodStuff, kind: String): Mono<Void> {
        if (random() > 0.8) {
            return Mono.error(RandomException())
        }
        val capacity: AtomicInteger? = this.stocks[stuff]!![kind]

        if (capacity == null) {
            return Mono.error(UnknownFoodStuffException(stuff, kind))
        } else if (capacity.decrementAndGet() < 0) {
            LOGGER.error("No more {}: {}", stuff, kind)
            return Mono.error(NoMoreFoodStuffException(stuff, kind))
        }
        return Mono.empty()
    }

}